src/utils.ts

```diff
@@ -823,3 +831,247 @@ export const messageSetPagination = <StreamChatGenerics extends ExtendableGeneri
     return messagePaginationLinear(params);
   }
 };
+
+/**
+ * A utility object used to prevent duplicate invocation of channel.watch() to be triggered when
+ * 'notification.message_new' and 'notification.added_to_channel' events arrive at the same time.
+ */
+const WATCH_QUERY_IN_PROGRESS_FOR_CHANNEL: Record<string, Promise<QueryChannelAPIResponse> | undefined> = {};
+
+type GetChannelParams<StreamChatGenerics extends ExtendableGenerics = DefaultGenerics> = {
+  client: StreamChat<StreamChatGenerics>;
+  channel?: Channel<StreamChatGenerics>;
+  id?: string;
+  members?: string[];
+  options?: ChannelQueryOptions<StreamChatGenerics>;
+  type?: string;
+};
+/**
+ * Calls channel.watch() if it was not already recently called. Waits for watch promise to resolve even if it was invoked previously.
+ * If the channel is not passed as a property, it will get it either by its channel.cid or by its members list and do the same.
+ * @param client
+ * @param members
+ * @param options
+ * @param type
+ * @param id
+ * @param channel
+ */
+export const getAndWatchChannel = async <StreamChatGenerics extends ExtendableGenerics = DefaultGenerics>({
+  channel,
+  client,
+  id,
+  members,
+  options,
+  type,
+}: GetChannelParams<StreamChatGenerics>) => {
+  if (!channel && !type) {
+    throw new Error('Channel or channel type have to be provided to query a channel.');
+  }
+
+  // unfortunately typescript is not able to infer that if (!channel && !type) === false, then channel or type has to be truthy
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const channelToWatch = channel || client.channel(type!, id, { members });
```

Comment:
TS can't infer because I feel you have a bug in the condition above. Shouldn't it be:

```ts
  if (!channel || !type) {
    throw new Error('Channel or channel type have to be provided to query a channel.');
  }
```

Reply:
I don't think so, because if a `channel` is passed and not a `type` then the `type` can be inferred from `channel.type`. It should be able to handle the situation of not passing a specific `channel` but passing its `type` so that later we can try to create a `cid` (either through `members` or `channel_id`).

The suggestion above would fail for mostly any invocation of this that does not have a `channel` passed to it.